libraries/AP_Follow/AP_Follow.cpp

```diff
@@ -0,0 +1,336 @@
+        Location loc_estimate{};
+        Vector3f vel_estimate;
+        UNUSED_RESULT(get_target_location_and_velocity(loc_estimate, vel_estimate));
+
+        // decode message
+        mavlink_global_position_int_t packet;
+        mavlink_msg_global_position_int_decode(&msg, &packet);
+
+        // ignore message if lat and lon are zero
+        if ((packet.lat != 0) || (packet.lon != 0)) {
+            return;
+        }
+
+        _target_location.lat = packet.lat;
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt / 10;     // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx * 0.01f; // velocity north
+        _target_velocity_ned.y = packet.vy * 0.01f; // velocity east
+        _target_velocity_ned.z = packet.vz * 0.01f; // velocity down
+        _last_location_update_ms = now;
+        if (packet.hdg <= 36000) {                  // heading (UINT16_MAX if unknown)
+            _target_heading = packet.hdg * 0.01f;   // convert centi-degrees to degrees
+            _last_heading_update_ms = now;
+        }
+        // initialise _sysid if zero to sender's id
+        if (_sysid_to_follow == 0) {
+            _sysid_to_follow = msg.sysid;
+        }
+        if ((AP_HAL::millis() - _last_location_sent_to_gcs > AP_GCS_INTERVAL_MS)) {
+            gcs().send_text(MAV_SEVERITY_INFO, "Foll: %u %ld %ld %4.2f\n",
+                            (unsigned)_sysid_to_follow,
+                            (long)_target_location.lat,
+                            (long)_target_location.lng,
+                            (double)(_target_location.alt * 0.01f));
+        }
+
+        // log lead's estimated vs reported position
+        DataFlash_Class::instance()->Log_Write("FOLL",
+                                               "TimeUS,Lat,Lon,Alt,VelX,VelY,VelZ,LatE,LonE,AltE",
+                                               "QLLifffLLi",
+                                               now,
+                                               _target_location.lat,
+                                               _target_location.lng,
+                                               _target_location.alt,
+                                               (double)_target_velocity_ned.x,
+                                               (double)_target_velocity_ned.y,
+                                               (double)_target_velocity_ned.z,
+                                               loc_estimate.lat,
+                                               loc_estimate.lng,
+                                               loc_estimate.alt
+                                               );
+    }
+}
+
+// get velocity estimate in NED frame
+bool AP_Follow::get_velocity_ned(Vector3f& vel, float dt) const
+{
+    vel = _target_velocity_ned + (_target_accel_ned * dt);
+    return true;
+}
+
+// initialise offsets to target if required
+void AP_Follow::init_offsets_if_required(const Vector3f& dist_vec_ned)
```

Comment:
Is this in m or cm ??

Reply:
it's in meters.  thanks, fixed in follow up commit.